configuration/configurator/ConfigSchemaHelper.cpp

```diff
@@ -0,0 +1,898 @@
+       s_pCConfigSchemaHelper->m_nTables = 0;
+    }
+    return s_pCConfigSchemaHelper;
+}
+
+CConfigSchemaHelper::CConfigSchemaHelper(const char* pBuildSetFile, const char* pBuildSetDir, const char* pDefaultDirOverride) : m_pBasePath(NULL), m_nTables(0),\
+    m_pEnvPropertyTree(NULL), m_pSchemaMapManager(NULL)
+{
+    assert(pBuildSetFile != NULL);
+    assert(pBuildSetDir != NULL);
+
+    CBuildSetManager::getInstance(pBuildSetFile, pBuildSetDir);
+    m_pSchemaMapManager = new CSchemaMapManager();
+}
+
+CConfigSchemaHelper::~CConfigSchemaHelper()
+{
+    delete[] m_pBasePath;
+    delete CConfigSchemaHelper::m_pSchemaMapManager;
+    CConfigSchemaHelper::m_pSchemaMapManager = NULL;
+    CConfigSchemaHelper::s_pCConfigSchemaHelper = NULL;
+}
+
+bool CConfigSchemaHelper::populateSchema()
+{
+    assert(m_pSchemaMapManager != NULL);
+
+    LOOP_THRU_BUILD_SET_MANAGER_BUILD_SET
+    {
+        const char *pSchemaName = CBuildSetManager::getInstance()->getBuildSetComponentFileName(idx);
+
+        if (pSchemaName != NULL)
+        {
+            CXSDNodeBase *pNull = NULL;
+            CSchema *pSchema = CSchema::load(pSchemaName, pNull);
+
+            assert(pSchema->getLinkCount() == 1);
+            m_pSchemaMapManager->setSchemaForXSD(pSchemaName, pSchema);
+        }
+    }
+    populateEnvXPath();
+
+    return true;
+}
+
+void CConfigSchemaHelper::printConfigSchema(StringBuffer &strXML) const
+{
+    assert(m_pSchemaMapManager != NULL);
+
+    const char *pComponent = NULL;
+    CSchema* pSchema = NULL;
+
+    LOOP_THRU_BUILD_SET_MANAGER_BUILD_SET
+    {
+        const char *pSchemaName = CBuildSetManager::getInstance()->getBuildSetComponentFileName(idx);
+
+        if (pComponent == NULL || strcmp(pComponent, pSchemaName) == 0)
+        {
+            const char* pXSDSchema = pSchemaName;
+
+            if (pXSDSchema == NULL)
+                continue;
+
+            pSchema = m_pSchemaMapManager->getSchemaForXSD(pSchemaName);
+
+            if (pSchema != NULL)
+            {
+                if (strXML.length() > 0 ? strcmp(strXML.str(), pXSDSchema) == 0 : true)
+                    pSchema->dump(::std::cout);
+            }
+        }
+    }
+}
+
+const char* CConfigSchemaHelper::printDocumentation(const char* comp)
+{
+    assert(comp != NULL && *comp != 0);
+    assert(m_pSchemaMapManager != NULL);
+
+    if (comp == NULL || *comp == 0)
+        return NULL;
+
+    CSchema* pSchema = NULL;
+
+    LOOP_THRU_BUILD_SET_MANAGER_BUILD_SET
+    {
+        const char *pSchemaName = CBuildSetManager::getInstance()->getBuildSetComponentFileName(idx);
+
+        if (pSchemaName != NULL && strcmp(comp, pSchemaName) == 0)
+        {
+             pSchema = m_pSchemaMapManager->getSchemaForXSD(pSchemaName);
+
+             assert(pSchema != NULL);
+
+             if (pSchema != NULL)
+             {
+                static StringBuffer strDoc;
```

Comment:
I'm not sure why this is static, but it's cleared every time we enter this section